src/script/interpreter.cpp

```diff
@@ -606,10 +606,29 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                     // To provide for future soft-fork extensibility, if the
                     // operand has the disabled lock-time flag set,
                     // CHECKSEQUENCEVERIFY behaves as a NOP.
-                    if ((nSequence & CTxIn::SEQUENCE_LOCKTIME_DISABLE_FLAG) != 0)
+                    // This also covers CTxIn::SEQUENCE_FINAL since all bits are
+                    // set in that case.
+                    if ((nSequence & CTxIn::SEQUENCE_LOCKTIME_DISABLE_FLAG) != 0) {
+                        // CHECKSEQUENCEVERIFY behaving as NOP, set error appropriately
+                        if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS)
```

Comment:
In 1bd2b15d82ebc80b3c30ab247287717a893cb37a

The approach of using `DISCOURAGE_UPGRADABLE_NOPS` within an already-upgraded NOP also feels very weird to me, especially as it requires loosening the unit tests to make them pass.